From: CloudFox-Kernel <kernel@cloudfox.dev>
Date: Mon, 3 Aug 2026 00:00:00 +0000
Subject: [PATCH 1/3] zsmalloc: add zs_lookup_class_index() for ZRAM
 recompression support

Add zs_lookup_class_index() function to zsmalloc, which returns the
index of the size_class that holds objects of the provided size.
This is needed by ZRAM recompression to determine if recompression
made progress (i.e., moved to a smaller size class).

Signed-off-by: CloudFox-Kernel <kernel@cloudfox.dev>
---
 mm/zsmalloc.c              | 19 +++++++++++++++++++
 include/linux/zsmalloc.h   |  1 +
 2 files changed, 20 insertions(+)

diff --git a/mm/zsmalloc.c b/mm/zsmalloc.c
index 9d1a19a..8f024e5 100644
--- a/mm/zsmalloc.c
+++ b/mm/zsmalloc.c
@@ -1231,6 +1231,25 @@ static bool zspage_full(struct size_class *class, struct zspage *zspage)
 	return get_zspage_inuse(zspage) == class->objs_per_zspage;
 }
 
+/**
+ * zs_lookup_class_index() - Returns index of the zsmalloc &size_class
+ * that hold objects of the provided size.
+ * @pool: zsmalloc pool to use
+ * @size: object size
+ *
+ * Context: Any context.
+ *
+ * Return: the index of the zsmalloc &size_class that hold objects of the
+ * provided size.
+ */
+unsigned int zs_lookup_class_index(struct zs_pool *pool, unsigned int size)
+{
+	struct size_class *class;
+
+	class = pool->size_class[get_size_class_index(size)];
+	return class->index;
+}
+EXPORT_SYMBOL_GPL(zs_lookup_class_index);
+
 unsigned long zs_get_total_pages(struct zs_pool *pool)
 {
 	return atomic_long_read(&pool->pages_allocated);
diff --git a/include/linux/zsmalloc.h b/include/linux/zsmalloc.h
index 60ccb83..e0b7a6a 100644
--- a/include/linux/zsmalloc.h
+++ b/include/linux/zsmalloc.h
@@ -55,6 +55,7 @@ void *zs_map_object(struct zs_pool *pool, unsigned long handle,
 void zs_unmap_object(struct zs_pool *pool, unsigned long handle);
 
 unsigned long zs_get_total_pages(struct zs_pool *pool);
+unsigned int zs_lookup_class_index(struct zs_pool *pool, unsigned int size);
 unsigned long zs_compact(struct zs_pool *pool);
 
 void zs_pool_stats(struct zs_pool *pool, struct zs_pool_stats *stats);